File: quality/run-qc.mjs
#!/usr/bin/env node
/**
 * §M-QC-RUNNER — meta-o's own `make qc` aggregator.
 *
 * Implements §A-AUTHORITATIVE-QC by eating the contract this project defines:
 * the gates come from the tracked manifest, a missing tool is a failure rather
 * than a skip, a mutated worktree invalidates the run, and the machine-readable
 * result goes to `$META_O_QC_RESULT` so nothing has to infer a pass from an
 * exit code.
 *
 * Dependency-free on purpose — a quality gate that cannot run because its own
 * dependency failed to install is a quality gate that gets disabled.
 */

import { execFileSync, spawnSync } from "node:child_process";
import { createHash } from "node:crypto";
import {
  closeSync,
  fsyncSync,
  mkdirSync,
  openSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

/** §M-QC-RUNNER — Repository root, resolved the way the digest defines it. */
const ROOT = execFileSync("git", ["rev-parse", "--show-toplevel"], {
  cwd: fileURLToPath(new URL(".", import.meta.url)),
  encoding: "utf8",
}).trim();

/** §M-QC-RUNNER — The porcelain listing of what the worktree has outstanding. */
function gitStatus() {
  return execFileSync("git", ["status", "--porcelain", "--untracked-files=all"], {
    cwd: ROOT,
    encoding: "utf8",
  });
}

/**
 * §M-QC-RUNNER — Snapshot the worktree by content, not by its status listing.
 *
 * The listing alone is not enough, and the Python template already says so:
 * `git status --porcelain` prints the same two characters and the same path
 * whether or not the bytes behind them changed, so a formatter rewriting a file
 * that was *already* dirty leaves the listing identical and the mutation check
 * passes — writing an attestable green result for content the gate itself
 * altered. A clean tree caught that and a dirty one did not, which is backwards.
 *
 * So the snapshot is the listing plus the bytes it refers to: the full diff
 * against `HEAD` for tracked changes, and a hash per untracked file.
 */
function worktreeFingerprint(listing) {
  const parts = [listing, execFileSync("git", ["diff", "HEAD"], { cwd: ROOT, encoding: "utf8" })];
  for (const line of listing.split("\n")) {
    if (!line.startsWith("?? ")) continue;
    const name = line.slice(3).trim().replace(/^"|"$/g, "");
    try {
      parts.push(`${name}:${createHash("sha256").update(readFileSync(join(ROOT, name))).digest("hex")}`);
    } catch (error) {
      // A file that vanished or became unreadable mid-run is itself a mutation.
      parts.push(`${name}:unreadable:${error.code ?? error.message}`);
    }
  }
  return parts.join("\n");
}

/**
 * §M-QC-RUNNER — Split a manifest command into argv.
 *
 * Handles the quoting the manifest actually uses and nothing more: a command
 * that needs a shell needs a script, because a shell here would also give a
 * gate the ability to chain, redirect and silently swallow a failure.
 */
function argv(command) {
  const parts = command.match(/"[^"]*"|'[^']*'|\S+/g) ?? [];
  return parts.map((part) =>
    (part.startsWith('"') && part.endsWith('"')) || (part.startsWith("'") && part.endsWith("'"))
      ? part.slice(1, -1)
      : part,
  );
}

/** §M-QC-RUNNER — Best-effort version of the tool a gate invokes. */
function toolVersion(executable) {
  const probe = spawnSync(executable, ["--version"], { encoding: "utf8", timeout: 30_000 });
  const text = `${probe.stdout ?? ""}${probe.stderr ?? ""}`.trim().split("\n")[0];
  return text || "unknown";
}

/**
 * §M-QC-RUNNER — Run one declared gate and classify its outcome.
 *
 * `ENOENT` is reported as a failure. That single decision separates a gate from
 * a decoration: the commonest way for one to stop working is for its tool to
 * disappear, and the most tempting response is to treat that as "not
 * applicable".
 */
function runGate(gate) {
  const command = String(gate.command);
  const [executable, ...args] = argv(command);
  const started = Date.now();

  const result = spawnSync(executable, args, { cwd: ROOT, stdio: "inherit" });

  if (result.error && result.error.code === "ENOENT") {
    return {
      id: gate.id,
      status: "failed",
      command,
      tool_version: "unknown",
      duration_ms: Date.now() - started,
      detail: `${executable} is not installed; a missing tool is a failure, not a skip`,
    };
  }
  if (result.error) throw result.error;

  return {
    id: gate.id,
    status: result.status === 0 ? "passed" : "failed",
    command,
    tool_version: toolVersion(executable),
    duration_ms: Date.now() - started,
  };
}

/**
 * §M-QC-RUNNER — Write the result atomically and durably.
 *
 * Rename alone survives a crashed process; it does not survive a lost machine,
 * and this file is the sole evidence behind one of the four attestations. The
 * run state is fsynced for the same reason, and the gate result had been left
 * out of that rule.
 */
function writeResult(path, value) {
  mkdirSync(dirname(path), { recursive: true });
  const temporary = `${path}.tmp`;
  rmSync(temporary, { force: true });
  const handle = openSync(temporary, "wx", 0o600);
  try {
    writeFileSync(handle, `${JSON.stringify(value, null, 2)}\n`);
    fsyncSync(handle);
  } finally {
    closeSync(handle);
  }
  renameSync(temporary, path);
}

/** §M-QC-RUNNER — Run every declared gate and report an attestable result. */
function main() {
  const manifestPath = join(ROOT, ".quality", "qc-manifest.json");
  const manifest = JSON.parse(readFileSync(manifestPath, "utf8"));

  const listing = gitStatus();
  const before = worktreeFingerprint(listing);
  const results = [];

  // Clean, not merely unchanged — but only when this run is producing an
  // attestation. An uncommitted edit that happens to fix a lint error makes
  // every gate pass while the digest in the result comes from HEAD, so the
  // attestation would describe content that was never the thing tested. A
  // developer running `make qc` by hand attests nothing, and failing them for a
  // dirty tree would just teach everyone to bypass the gate.
  const attesting = Boolean(process.env["META_O_QC_RESULT"]);
  if (listing.trim() !== "" && attesting) {
    process.stderr.write(`the worktree is dirty; qc cannot attest content that is not committed:\n${listing}`);
    results.push({
      id: "clean-worktree",
      status: "failed",
      command: "git status --porcelain --untracked-files=all",
      tool_version: "git",
      duration_ms: 0,
    });
  }

  for (const gate of manifest.gates) {
    if (gate.policy === "not_applicable") {
      process.stdout.write(`== ${gate.id}: not applicable (${gate.rationale ?? ""})\n`);
      results.push({
        id: gate.id,
        status: "not_applicable",
        command: String(gate.command),
        tool_version: "n/a",
        duration_ms: 0,
      });
      continue;
    }
    process.stdout.write(`== ${gate.id}: ${gate.command}\n`);
    results.push(runGate(gate));
  }

  if (worktreeFingerprint(gitStatus()) !== before) {
    process.stderr.write("qc mutated the worktree; a gate that rewrites the tree cannot attest it\n");
    results.push({
      id: "non-mutating",
      status: "failed",
      command: "git status --porcelain --untracked-files=all + git diff HEAD",
      tool_version: "git",
      duration_ms: 0,
    });
  }

  const destination = process.env["META_O_QC_RESULT"];
  if (destination) {
    writeResult(destination, {
      schema_version: 1,
      snapshot_digest: process.env["META_O_SNAPSHOT_DIGEST"] ?? "",
      gates: results,
    });
    process.stdout.write(`qc result written to ${destination}\n`);
  }

  const failed = results.filter((gate) => gate.status === "failed").map((gate) => gate.id);
  if (failed.length > 0) {
    process.stderr.write(`\nqc FAILED: ${failed.join(", ")}\n`);
    return 1;
  }
  process.stdout.write("\nqc passed\n");
  return 0;
}

process.exitCode = main();

File: templates/python/quality/run_qc.py
#!/usr/bin/env python3
"""§M-QC-RUNNER — The authoritative `make qc` aggregator.

Implements the project side of the QC contract. Its job is not to run commands —
a shell can do that — but to make the three silent failures impossible: a tool
that was missing and got skipped, a gate that never ran but was assumed to pass,
and a formatter that quietly rewrote the tree it was supposed to be judging.

Writes a machine-readable result to `$META_O_QC_RESULT` when the orchestrator
set it. A missing result is never a pass on the other side, so writing it
atomically matters as much as computing it.
"""

from __future__ import annotations

import json
import os
import hashlib
import shlex
import shutil
import subprocess
import sys
import time
from pathlib import Path
from typing import Any

from _common import project_root, read_json, write_json

SCHEMA_VERSION = 1


def git_status(root: Path) -> str:
    """§M-QC-RUNNER — Snapshot the worktree by content, not by its status listing.

    The listing alone was not enough. `git status --porcelain` prints the same
    two characters and the same path whether or not the file behind them
    changed, so a formatter rewriting a file that was *already* dirty left the
    text identical and the mutation check passed — and the run wrote an
    attestable green result for content the gate itself had altered. A clean
    tree caught it and a dirty one did not, which is backwards: a dirty tree is
    where a stray rewrite is hardest to notice.

    So the snapshot is the listing plus the bytes it refers to: the full diff
    against `HEAD` for tracked changes, and a hash per untracked file.
    """
    listing = _git(root, ["status", "--porcelain", "--untracked-files=all"])
    parts = [listing, _git(root, ["diff", "HEAD"])]
    for line in listing.splitlines():
        if not line.startswith("?? "):
            continue
        path = root / line[3:].strip().strip('"')
        try:
            parts.append(f"{line[3:]}:{hashlib.sha256(path.read_bytes()).hexdigest()}")
        except OSError as error:  # a file that vanished mid-run is itself a mutation
            parts.append(f"{line[3:]}:unreadable:{error}")
    return "\n".join(parts)


def _git(root: Path, args: list[str]) -> str:
    """§M-QC-RUNNER — Run one read-only git command and return its stdout."""
    result = subprocess.run(
        ["git", *args],
        cwd=root,
        capture_output=True,
        text=True,
        check=False,
    )
    return result.stdout


def tool_version(command: str) -> str:
    """§M-QC-RUNNER — Best-effort version of the tool a gate invokes.

    Recorded because "the lint gate passed" means nothing without knowing which
    linter said so; a version bump that silently drops a rule is otherwise
    invisible in the result.
    """
    parts = shlex.split(command)
    if not parts:
        return "unknown"
    executable = parts[0]
    if executable in {"python", "python3"} and len(parts) > 2 and parts[1] == "-m":
        executable = f"{executable} -m {parts[2]}"
    try:
        result = subprocess.run(
            [*shlex.split(executable), "--version"],
            capture_output=True,
            text=True,
            check=False,
            timeout=30,
        )
        text = (result.stdout or result.stderr).strip().splitlines()
        return text[0] if text else "unknown"
    except (OSError, subprocess.SubprocessError):
        return "unknown"


def run_gate(gate: dict[str, Any], root: Path) -> dict[str, Any]:
    """§M-QC-RUNNER — Execute one declared gate and classify its outcome.

    A missing executable is a failure, never a skip. That single decision is the
    difference between a quality gate and a decorative one: the most common way
    for a gate to stop working is for its tool to disappear from the
    environment, and the most tempting way to handle that is to shrug.
    """
    identifier = str(gate["id"])
    command = str(gate["command"])
    started = time.monotonic()

    parts = shlex.split(command)
    if not parts:
        return {
            "id": identifier,
            "status": "failed",
            "command": command,
            "tool_version": "unknown",
            "duration_ms": 0,
            "detail": "empty command",
        }

    if shutil.which(parts[0]) is None:
        return {
            "id": identifier,
            "status": "failed",
            "command": command,
            "tool_version": "unknown",
            "duration_ms": 0,
            "detail": f"tool {parts[0]} is not installed; a missing tool is a failure, not a skip",
        }

    environment = dict(os.environ)
    environment["PYTHONPATH"] = os.pathsep.join(
        [str(root / "quality"), environment.get("PYTHONPATH", "")]
    ).rstrip(os.pathsep)

    result = subprocess.run(parts, cwd=root, env=environment, check=False)
    duration = int((time.monotonic() - started) * 1000)

    return {
        "id": identifier,
        "status": "passed" if result.returncode == 0 else "failed",
        "command": command,
        "tool_version": tool_version(command),
        "duration_ms": duration,
    }


def main() -> int:
    """§M-QC-RUNNER — Run every declared gate and write the attestable result."""
    root = project_root()
    manifest_path = root / ".quality" / "qc-manifest.json"
    manifest = read_json(manifest_path)
    if manifest is None:
        sys.stderr.write(f"{manifest_path} is missing; the QC contract is not declared\n")
        return 1

    before = git_status(root)
    results: list[dict[str, Any]] = []

    for gate in manifest.get("gates", []):
        identifier = str(gate.get("id", "?"))
        if gate.get("policy") == "not_applicable":
            sys.stdout.write(f"== {identifier}: not applicable ({gate.get('rationale', '')})\n")
            results.append(
                {
                    "id": identifier,
                    "status": "not_applicable",
                    "command": str(gate.get("command", "")),
                    "tool_version": "n/a",
                    "duration_ms": 0,
                }
            )
            continue

        sys.stdout.write(f"== {identifier}: {gate.get('command')}\n")
        sys.stdout.flush()
        results.append(run_gate(gate, root))

    after = git_status(root)
    mutated = before != after
    if mutated:
        sys.stderr.write(
            "qc mutated the worktree; a gate that rewrites the tree cannot attest it\n"
        )
        results.append(
            {
                "id": "non-mutating",
                "status": "failed",
                "command": "git status --porcelain --untracked-files=all + git diff HEAD",
                "tool_version": "git",
                "duration_ms": 0,
            }
        )

    destination = os.environ.get("META_O_QC_RESULT")
    if destination:
        write_json(
            Path(destination),
            {
                "schema_version": SCHEMA_VERSION,
                "snapshot_digest": os.environ.get("META_O_SNAPSHOT_DIGEST", ""),
                "gates": results,
            },
        )
        sys.stdout.write(f"qc result written to {destination}\n")
    else:
        sys.stdout.write(f"{json.dumps({'gates': results}, indent=2)}\n")

    failed = [gate["id"] for gate in results if gate["status"] == "failed"]
    if failed:
        sys.stderr.write(f"qc FAILED: {', '.join(failed)}\n")
        return 1
    sys.stdout.write("qc passed\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
